### PR TITLE
Speed up deleting relationship by parent. (Speeds up publishing/ saving)

### PR DESCRIPTION
**Pull request by nzdev**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes N/A

### Description
What?
Improve query for Delete relationship by parent.

Why?
Faster deletion of relationships = reduced time to publish.

How to test?
Existing tests pass. Existing relationships on existing content are removed when the content is saved.
Execution time should be reduced on sql server/ sql azure

### CODE REVIEW

File: src/Umbraco.Core/Constants-SqlTemplates.cs [hunk 1]
```diff
@@ -14,7 +14,11 @@ public static class VersionableRepository
                 public const string GetParentNode = "Umbraco.Core.VersionableRepository.GetParentNode";
                 public const string GetReservedId = "Umbraco.Core.VersionableRepository.GetReservedId";
             }
-
+            public static class RelationRepository
+            {
+                public const string DeleteByParentAll = "Umbraco.Core.RelationRepository.DeleteByParent";
```
Comment (Shazwazza) [inline src/Umbraco.Core/Constants-SqlTemplates.cs:19]: I can see these were in use before but that has been reverted. Any reason why we still can't use Sql templates? Even if it's just for the non-sqlce?

**Review by Shazwazza — CHANGES_REQUESTED** [review]
Yeah SqlCe is a bugger. There's probably quite a few queries that can be easily optimized for Sql server. Some of which could be wrapped possibly on ext methods to the SqlSyntaxProvider. Do you have any stats (i.e. Sql execution plan) of running the orig queries vs. the pure delete with inner join query? There was one failing test for some reason but don't think it's related to this, i've just re-run the build to see what happens.

File: src/Umbraco.Core/Constants-SqlTemplates.cs (hunk 1 above)
Comment (nzdev) [reply src/Umbraco.Core/Constants-SqlTemplates.cs:19]: I did set up the template, but I didn't know enough about how they worked with two sets of parameters so I reverted the change. Hopefully I can get some assistance with this. I think it went from ~1000ms to <40ms. But I'm not sure. There are too many queries I've looked at lately to remember. But the difference was huge.

Comment (nzdev) [reply src/Umbraco.Core/Constants-SqlTemplates.cs:19]: Original
DELETE [umbracoRelation] 
where id in ( 
SELECT r.id
  FROM [dbo].[umbracoRelation] r
  INNER JOIN  umbracoRelationType rt on rt.id = r.relType
  where r.parentId = 155928)

image

New:
  DELETE umbracoRelation
  FROM umbracoRelation
  INNER JOIN umbracoRelationType rt on rt.id = relType
  where umbracoRelation.parentId = parentId

image

**Comment by Shazwazza** [comment]
> I did set up the template, but I didn't know enough about how they worked with two sets of parameters so I reverted the change. Hopefully I can get some assistance with this

ok cool no worries, i can have a look when merged too. I haven't looked at that for a while so not sure off the top of my head.

For the query comparisons - if you put both queries in the same query window and execute, it will show results from both queries along with side by side results of the execution plan and it will tell you the "% cost relative to the batch" so you can see exactly by how much % one query is over the other.

**Comment by Shazwazza** [comment]
I get this result 
image

So shows the newer version is definitely better 🎉

**Comment by Shazwazza** [comment]
I'll get this merged and have a look at the SQL templates. Great stuff :)

**Comment by Shazwazza** [comment]
@nzdev here's an update which now uses Sql templates for these https://github.com/umbraco/Umbraco-CMS/commit/9269271649b9b2f6296fb09b6893b3d27c0f94bf

**Comment by nzdev** [comment]
@Shazwazza have a look at https://github.com/umbraco/Umbraco-CMS/pull/9508/commits/39b1de4a98c5d009bb607370fb6dc2590698b1aa . I think the issue was argIn<string> didn't work

**Comment by Shazwazza** [comment]
@nzdev strange, i've just tested on my machine and it executes just fine, but that commit you've referenced also wasn't checking for sqlce so the build may have been failing because of that

**Comment by nzdev** [comment]
@Shazwazza Possibly, though I checked the query in the sql profiler at the time and it was not creating the in clause for multiple alias's properly.

**Comment by Shazwazza** [comment]
This is the resulting sql string

```
DELETE [umbracoRelation]
FROM [umbracoRelation]
INNER JOIN [umbracoRelationType]
ON [umbracoRelation].[relType] = [umbracoRelationType].[id]
WHERE (([umbracoRelation].[parentId] = @0))
AND ([umbracoRelationType].[alias] IN (@1,@2))
```

with these args

image

**Comment by nzdev** [comment]
looks correct to me 👍

**Pull request merged**